// @progress/kendo-react-scheduler/dist/npm/views/week/WeekView.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultProps = exports.WeekView = void 0;
var React = require("react");
var MultiDayView_1 = require("../day/MultiDayView");
var constants_1 = require("../../constants");
var kendo_date_math_1 = require("@progress/kendo-date-math");
var messages_1 = require("../../messages");
var utils_1 = require("../../utils");
/**
 * Represents the KendoReact [WeekView]({% slug week_views_scheduler %}) Component.
 */
var WeekView = function (props) { return (React.createElement(MultiDayView_1.MultiDayView, __assign({}, props, { step: constants_1.DAYS_IN_WEEK_COUNT, numberOfDays: props.numberOfDays }))); };
exports.WeekView = WeekView;
var GET_START_DATE = function (date) { return (0, kendo_date_math_1.getDate)(date); };
var GET_END_DATE = function (start, numberOfDays) { return (0, kendo_date_math_1.getDate)((0, kendo_date_math_1.addDays)(start, numberOfDays || 1)); };
var weekViewDateRange = function (_a) {
    var intl = _a.intl, date = _a.date, timezone = _a.timezone;
    var normalized = kendo_date_math_1.ZonedDate.fromLocalDate(date, timezone);
    var firstDate = GET_START_DATE((0, kendo_date_math_1.firstDayInWeek)(normalized, intl.firstDay()));
    var lastDate = GET_END_DATE(firstDate, constants_1.DAYS_IN_WEEK_COUNT);
    var zonedStart = kendo_date_math_1.ZonedDate.fromUTCDate((0, utils_1.toUTCDateTime)(firstDate), timezone);
    var zonedEnd = kendo_date_math_1.ZonedDate.fromUTCDate((0, utils_1.toUTCDateTime)(lastDate), timezone);
    var start = new Date(zonedStart.getTime());
    var end = new Date(zonedEnd.getTime());
    return {
        start: start,
        end: end,
        zonedStart: zonedStart,
        zonedEnd: zonedEnd
    };
};
/** @hidden */
exports.defaultProps = {
    name: 'week',
    slotDuration: 60,
    slotDivisions: 2,
    numberOfDays: constants_1.DAYS_IN_WEEK_COUNT,
    dateRange: weekViewDateRange,
    title: function (localization) { return localization.toLanguageString(messages_1.weekViewTitle, messages_1.messages[messages_1.weekViewTitle]); },
    selectedDateFormat: '{0:D} - {1:D}',
    selectedShortDateFormat: '{0:d} - {1:d}'
};
exports.WeekView.defaultProps = exports.defaultProps;
exports.WeekView.displayName = 'KendoReactSchedulerWeekView';
