// @progress/kendo-react-scheduler/dist/npm/views/day/DayView.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DayView = void 0;
var React = require("react");
var MultiDayView_1 = require("../day/MultiDayView");
var messages_1 = require("../../messages");
/**
 * Represents the KendoReact [DayView]({% slug day_view_scheduler %}) Component.
 */
var DayView = function (props) { return React.createElement(MultiDayView_1.MultiDayView, __assign({}, props)); };
exports.DayView = DayView;
var defaultProps = {
    name: 'day',
    title: function (localization) { return localization.toLanguageString(messages_1.dayViewTitle, messages_1.messages[messages_1.dayViewTitle]); },
    dateRange: MultiDayView_1.defaultProps.dateRange,
    step: 1,
    numberOfDays: 1,
    slotDuration: 60,
    slotDivisions: 2,
    selectedDateFormat: '{0:D}',
    selectedShortDateFormat: '{0:d}'
};
exports.DayView.defaultProps = defaultProps;
exports.DayView.displayName = 'KendoReactSchedulerDayView';
