// @progress/kendo-react-scheduler/dist/npm/context/SchedulerEditTaskContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchedulerEditTaskContext = exports.useSchedulerEditTaskShowOccurrenceDialogContext = exports.SchedulerEditTaskShowOccurrenceDialogContext = exports.useSchedulerEditTaskShowRemoveDialogContext = exports.SchedulerEditTaskShowRemoveDialogContext = exports.useSchedulerEditTaskRemoveItemContext = exports.SchedulerEditTaskRemoveItemContext = void 0;
var React = require("react");
/** @hidden */
exports.SchedulerEditTaskRemoveItemContext = React.createContext([null]);
/** @hidden */
var useSchedulerEditTaskRemoveItemContext = function () { return React.useContext(exports.SchedulerEditTaskRemoveItemContext); };
exports.useSchedulerEditTaskRemoveItemContext = useSchedulerEditTaskRemoveItemContext;
/** @hidden */
exports.SchedulerEditTaskShowRemoveDialogContext = React.createContext([false]);
/** @hidden */
var useSchedulerEditTaskShowRemoveDialogContext = function () { return React.useContext(exports.SchedulerEditTaskShowRemoveDialogContext); };
exports.useSchedulerEditTaskShowRemoveDialogContext = useSchedulerEditTaskShowRemoveDialogContext;
/** @hidden */
exports.SchedulerEditTaskShowOccurrenceDialogContext = React.createContext([false]);
/** @hidden */
var useSchedulerEditTaskShowOccurrenceDialogContext = function () { return React.useContext(exports.SchedulerEditTaskShowOccurrenceDialogContext); };
exports.useSchedulerEditTaskShowOccurrenceDialogContext = useSchedulerEditTaskShowOccurrenceDialogContext;
/** @hidden */
var SchedulerEditTaskContext = function (props) {
    return (React.createElement(exports.SchedulerEditTaskRemoveItemContext.Provider, { value: props.remove },
        React.createElement(exports.SchedulerEditTaskShowRemoveDialogContext.Provider, { value: props.showRemoveDialog },
            React.createElement(exports.SchedulerEditTaskShowOccurrenceDialogContext.Provider, { value: props.showOccurrenceDialog }, props.children))));
};
exports.SchedulerEditTaskContext = SchedulerEditTaskContext;
