// @progress/kendo-react-scheduler/dist/es/services/occurrenceService.js
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { getter, clone } from '@progress/kendo-react-common';
import { parseRule, expand } from '@progress/kendo-recurrence';
import { ZonedDate } from '@progress/kendo-date-math';
import { setField, isMaster } from '../utils';
/** @hidden */
export var toOccurrences = function (data, _a) {
    var dateRange = _a.dateRange, fields = _a.fields, timezone = _a.timezone;
    var result = data
        .map(readItem(fields))
        .reduce(occurrenceReducer(dateRange, timezone, fields), [])
        .filter(function (occurrence) { return !isMaster(occurrence.dataItem, fields); });
    return result;
};
var readItem = function (fields) { return function (dataItem) {
    var result = {
        uid: getter(fields.id || 'id')(dataItem),
        start: getter(fields.start || 'start')(dataItem),
        startTimezone: getter(fields.startTimezone || 'startTimezone')(dataItem),
        originalStart: getter(fields.originalStart || 'originalStart')(dataItem),
        end: getter(fields.end || 'end')(dataItem),
        endTimezone: getter(fields.endTimezone || 'endTimezone')(dataItem),
        isAllDay: getter(fields.isAllDay || 'isAllDay')(dataItem),
        title: getter(fields.title || 'title')(dataItem),
        description: getter(fields.description || 'description')(dataItem),
        occurrenceId: getter('occurrenceId')(dataItem),
        recurrenceRule: getter(fields.recurrenceRule || 'recurrenceRule')(dataItem),
        recurrenceExceptions: getter(fields.recurrenceExceptions || 'recurrenceExceptions')(dataItem),
        recurrenceId: getter(fields.recurrenceId || 'recurrenceId')(dataItem),
        dataItem: clone(dataItem)
    };
    return result;
}; };
var occurrenceReducer = function (dateRange, timezone, fields) { return function (acc, current) {
    return __spreadArray(__spreadArray([], acc, true), ((Boolean(current.recurrenceRule)
        && Boolean(current.recurrenceId === null || current.recurrenceId === undefined)
        && isMaster(current.dataItem, fields))
        ? __spreadArray([], occurrences(current, { dateRange: dateRange, timezone: timezone, fields: fields }), true) : [current]), true);
}; };
var occurrences = function (item, _a) {
    var dateRange = _a.dateRange, timezone = _a.timezone, fields = _a.fields;
    var rrule = item.recurrenceRule;
    var rule = parseRule({ recurrenceRule: rrule });
    // changed as for display purposes timezone of the scheduler is the correct one
    if (!rule.start) {
        rule.start = ZonedDate.fromLocalDate(item.start, timezone);
    }
    if (!rule.end) {
        rule.end = ZonedDate.fromLocalDate(item.end, timezone);
    }
    var exceptionRule = item.recurrenceExceptions;
    if (exceptionRule) {
        rule.exceptionDates = exceptionRule
            .map(function (exDate) {
            return ZonedDate.fromLocalDate(exDate, timezone);
        });
    }
    var utcRangeStart = dateRange.zonedStart;
    var utcRangeEnd = dateRange.zonedEnd;
    var series = expand(rule, {
        rangeStart: utcRangeStart,
        rangeEnd: utcRangeEnd
    });
    if (!series.events.length) {
        return [];
    }
    var expanded = series.events.map(function (occurrence, idx) {
        var occurrenceItem = clone(item);
        var occurrenceDataItem = clone(item.dataItem);
        occurrenceItem.recurrenceId = occurrenceItem.uid;
        setField(occurrenceDataItem, fields.recurrenceId, item.uid);
        occurrenceItem.originalStart = occurrence.start.toLocalDate();
        setField(occurrenceDataItem, fields.originalStart, occurrence.start.toLocalDate());
        occurrenceItem.start = occurrence.start.toLocalDate();
        setField(occurrenceDataItem, fields.start, occurrence.start.toLocalDate());
        occurrenceItem.end = occurrence.end.toLocalDate();
        setField(occurrenceDataItem, fields.end, occurrence.end.toLocalDate());
        occurrenceItem.occurrenceId = String(idx);
        setField(occurrenceDataItem, 'occurrenceId', String(idx));
        occurrenceItem.dataItem = occurrenceDataItem;
        return occurrenceItem;
    });
    return expanded;
};
