// @progress/kendo-react-scheduler/dist/npm/messages/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.recurrenceEditorOffsetPositionsFirst = exports.recurrenceEditorFrequenciesYearly = exports.recurrenceEditorFrequenciesWeekly = exports.recurrenceEditorFrequenciesNever = exports.recurrenceEditorFrequenciesMonthly = exports.recurrenceEditorFrequenciesDaily = exports.recurrenceEditorYearlyInterval = exports.recurrenceEditorYearlyRepeatOn = exports.recurrenceEditorYearlyRepeatEvery = exports.recurrenceEditorYearlyOf = exports.recurrenceEditorMonthlyRepeatOn = exports.recurrenceEditorMonthlyRepeatEvery = exports.recurrenceEditorMonthlyInterval = exports.recurrenceEditorMonthlyDay = exports.recurrenceEditorWeeklyRepeatOn = exports.recurrenceEditorWeeklyRepeatEvery = exports.recurrenceEditorWeeklyInterval = exports.recurrenceEditorDailyRepeatEvery = exports.recurrenceEditorDailyInterval = exports.recurrenceEditorRepeat = exports.editorEventTimeZone = exports.editorEventSeparateTimeZones = exports.editorEventDescription = exports.editorEventAllDay = exports.editorEventEndTimeZone = exports.editorEventEnd = exports.editorEventStartTimeZone = exports.editorEventStart = exports.editorEventTitle = exports.editorTitle = exports.editorCancel = exports.editorSave = exports.editorRecurringDialogTitle = exports.editorRecurringConfirmation = exports.editorSeries = exports.editorOccurrence = exports.showWorkDay = exports.showFullDay = exports.nextTitle = exports.previousTitle = exports.deleteTitle = exports.monthViewTitle = exports.dayViewTitle = exports.agendaViewTitle = exports.workWeekViewTitle = exports.weekViewTitle = exports.timelineViewTitle = exports.today = exports.allDay = exports.allEvents = void 0;
exports.messages = exports.moreEvents = exports.noEvents = exports.editorValidationEnd = exports.editorValidationStart = exports.editorValidationRequired = exports.eventTitle = exports.timeTitle = exports.dateTitle = exports.deleteRecurringDialogTitle = exports.deleteDialogTitle = exports.deleteSeries = exports.deleteOccurrence = exports.deleteRecurringConfirmation = exports.deleteConfirmation = exports.editorDelete = exports.recurrenceEditorEndOn = exports.recurrenceEditorEndNever = exports.recurrenceEditorEndLabel = exports.recurrenceEditorEndOccurrence = exports.recurrenceEditorEndAfter = exports.recurrenceEditorWeekdaysWeekendday = exports.recurrenceEditorWeekdaysWeekday = exports.recurrenceEditorWeekdaysDay = exports.recurrenceEditorOffsetPositionsLast = exports.recurrenceEditorOffsetPositionsFourth = exports.recurrenceEditorOffsetPositionsThird = exports.recurrenceEditorOffsetPositionsSecond = void 0;
/** @hidden */
exports.allEvents = 'scheduler.allEvents';
/** @hidden */
exports.allDay = 'scheduler.allDay';
/** @hidden */
exports.today = 'scheduler.today';
/** @hidden */
exports.timelineViewTitle = 'scheduler.timelineViewTitle';
/** @hidden */
exports.weekViewTitle = 'scheduler.weekViewTitle';
/** @hidden */
exports.workWeekViewTitle = 'scheduler.workWeekViewTitle';
/** @hidden */
exports.agendaViewTitle = 'scheduler.agendaViewTitle';
/** @hidden */
exports.dayViewTitle = 'scheduler.dayViewTitle';
/** @hidden */
exports.monthViewTitle = 'scheduler.monthViewTitle';
/** @hidden */
exports.deleteTitle = 'scheduler.deleteTitle';
/** @hidden */
exports.previousTitle = 'scheduler.previousTitle';
/** @hidden */
exports.nextTitle = 'scheduler.nextTitle';
/** @hidden */
exports.showFullDay = 'scheduler.showFullDay';
/** @hidden */
exports.showWorkDay = 'scheduler.showWorkDay';
/** @hidden */
exports.editorOccurrence = 'scheduler.editOccurrence';
/** @hidden */
exports.editorSeries = 'scheduler.editSeries';
/** @hidden */
exports.editorRecurringConfirmation = 'scheduler.editRecurringConfirmation';
/** @hidden */
exports.editorRecurringDialogTitle = 'scheduler.editRecurringDialogTitle';
/** @hidden */
exports.editorSave = 'scheduler.editSave';
/** @hidden */
exports.editorCancel = 'scheduler.editCancel';
/** @hidden */
exports.editorTitle = 'scheduler.editorTitle';
/** @hidden */
exports.editorEventTitle = 'scheduler.editorEventTitle';
/** @hidden */
exports.editorEventStart = 'scheduler.editorEventStart';
/** @hidden */
exports.editorEventStartTimeZone = 'scheduler.editorEventStartTimeZone';
/** @hidden */
exports.editorEventEnd = 'scheduler.editorEventEnd';
/** @hidden */
exports.editorEventEndTimeZone = 'scheduler.editorEventEndTimeZone';
/** @hidden */
exports.editorEventAllDay = 'scheduler.editorEventAllDay';
/** @hidden */
exports.editorEventDescription = 'scheduler.editorEventDescription';
/** @hidden */
exports.editorEventSeparateTimeZones = 'scheduler.editorEventSeparateTimeZones';
/** @hidden */
exports.editorEventTimeZone = 'scheduler.editorEventTimeZone';
/** @hidden */
exports.recurrenceEditorRepeat = 'scheduler.recurrenceEditorRepeat';
/** @hidden */
exports.recurrenceEditorDailyInterval = 'scheduler.recurrenceEditorDailyInterval';
/** @hidden */
exports.recurrenceEditorDailyRepeatEvery = 'scheduler.recurrenceEditorDailyRepeatEvery';
/** @hidden */
exports.recurrenceEditorWeeklyInterval = 'scheduler.recurrenceEditorWeeklyInterval';
/** @hidden */
exports.recurrenceEditorWeeklyRepeatEvery = 'scheduler.recurrenceEditorWeeklyRepeatEvery';
/** @hidden */
exports.recurrenceEditorWeeklyRepeatOn = 'scheduler.recurrenceEditorWeeklyRepeatOn';
/** @hidden */
exports.recurrenceEditorMonthlyDay = 'scheduler.recurrenceEditorMonthlyDay';
/** @hidden */
exports.recurrenceEditorMonthlyInterval = 'scheduler.recurrenceEditorMonthlyInterval';
/** @hidden */
exports.recurrenceEditorMonthlyRepeatEvery = 'scheduler.recurrenceEditorMonthlyRepeatEvery';
/** @hidden */
exports.recurrenceEditorMonthlyRepeatOn = 'scheduler.recurrenceEditorMonthlyRepeatOn';
/** @hidden */
exports.recurrenceEditorYearlyOf = 'scheduler.recurrenceEditorYearlyOf';
/** @hidden */
exports.recurrenceEditorYearlyRepeatEvery = 'scheduler.recurrenceEditorYearlyRepeatEvery';
/** @hidden */
exports.recurrenceEditorYearlyRepeatOn = 'scheduler.recurrenceEditorYearlyRepeatOn';
/** @hidden */
exports.recurrenceEditorYearlyInterval = 'scheduler.recurrenceEditorYearlyInterval';
/** @hidden */
exports.recurrenceEditorFrequenciesDaily = 'scheduler.recurrenceEditorFrequenciesDaily';
/** @hidden */
exports.recurrenceEditorFrequenciesMonthly = 'scheduler.recurrenceEditorFrequenciesMonthly';
/**  @hidden */
exports.recurrenceEditorFrequenciesNever = 'scheduler.recurrenceEditorFrequenciesNever';
/** @hidden */
exports.recurrenceEditorFrequenciesWeekly = 'scheduler.recurrenceEditorFrequenciesWeekly';
/** @hidden */
exports.recurrenceEditorFrequenciesYearly = 'scheduler.recurrenceEditorFrequenciesYearly';
/** @hidden */
exports.recurrenceEditorOffsetPositionsFirst = 'scheduler.recurrenceEditorOffsetPositionsFirst';
/** @hidden */
exports.recurrenceEditorOffsetPositionsSecond = 'scheduler.recurrenceEditorOffsetPositionsSecond';
/** @hidden */
exports.recurrenceEditorOffsetPositionsThird = 'scheduler.recurrenceEditorOffsetPositionsThird';
/** @hidden */
exports.recurrenceEditorOffsetPositionsFourth = 'scheduler.recurrenceEditorOffsetPositionsFourth';
/** @hidden */
exports.recurrenceEditorOffsetPositionsLast = 'scheduler.recurrenceEditorOffsetPositionsLast';
/** @hidden */
exports.recurrenceEditorWeekdaysDay = 'scheduler.recurrenceEditorWeekdaysDay';
/** @hidden */
exports.recurrenceEditorWeekdaysWeekday = 'scheduler.recurrenceEditorWeekdaysWeekday';
/** @hidden */
exports.recurrenceEditorWeekdaysWeekendday = 'scheduler.recurrenceEditorWeekdaysWeekendday';
/** @hidden */
exports.recurrenceEditorEndAfter = 'scheduler.recurrenceEditorEndAfter';
/** @hidden */
exports.recurrenceEditorEndOccurrence = 'scheduler.recurrenceEditorEndOccurrence';
/** @hidden */
exports.recurrenceEditorEndLabel = 'scheduler.recurrenceEditorEndLabel';
/** @hidden */
exports.recurrenceEditorEndNever = 'scheduler.recurrenceEditorEndNever';
/** @hidden */
exports.recurrenceEditorEndOn = 'scheduler.recurrenceEditorEndOn';
/** @hidden */
exports.editorDelete = 'scheduler.editorDelete';
/** @hidden */
exports.deleteConfirmation = 'scheduler.deleteConfirmation';
/** @hidden */
exports.deleteRecurringConfirmation = 'scheduler.deleteRecurringConfirmation';
/** @hidden */
exports.deleteOccurrence = 'scheduler.deleteOccurrence';
/** @hidden */
exports.deleteSeries = 'scheduler.deleteSeries';
/** @hidden */
exports.deleteDialogTitle = 'scheduler.deleteDialogTitle';
/** @hidden */
exports.deleteRecurringDialogTitle = 'scheduler.deleteRecurringDialogTitle';
/** @hidden */
exports.dateTitle = 'scheduler.dateTitle';
/** @hidden */
exports.timeTitle = 'scheduler.timeTitle';
/** @hidden */
exports.eventTitle = 'scheduler.eventTitle';
/** @hidden */
exports.editorValidationRequired = 'scheduler.editorValidationRequired';
/** @hidden */
exports.editorValidationStart = 'scheduler.editorValidationStart';
/** @hidden */
exports.editorValidationEnd = 'scheduler.editorValidationEnd';
/** @hidden */
exports.noEvents = 'scheduler.noEvents';
/** @hidden */
exports.moreEvents = 'scheduler.moreEvents';
/**
 * @hidden
 */
exports.messages = (_a = {},
    _a[exports.editorValidationRequired] = 'Field is required.',
    _a[exports.editorValidationStart] = 'Start time must be be before End time.',
    _a[exports.editorValidationEnd] = 'End time must be after Start time.',
    _a[exports.allEvents] = 'All events',
    _a[exports.allDay] = 'all day',
    _a[exports.today] = 'Today',
    _a[exports.dayViewTitle] = 'Day',
    _a[exports.monthViewTitle] = 'Month',
    _a[exports.weekViewTitle] = 'Week',
    _a[exports.workWeekViewTitle] = 'Work Week',
    _a[exports.timelineViewTitle] = 'Timeline',
    _a[exports.agendaViewTitle] = 'Agenda',
    _a[exports.deleteTitle] = 'Delete',
    _a[exports.previousTitle] = 'Previous',
    _a[exports.nextTitle] = 'Next',
    _a[exports.showFullDay] = 'Show full day',
    _a[exports.showWorkDay] = 'Show business hours',
    _a[exports.editorOccurrence] = 'Edit current occurrence',
    _a[exports.editorSeries] = 'Edit the series',
    _a[exports.editorRecurringConfirmation] = 'Do you want to edit only this event occurrence or the whole series?',
    _a[exports.editorRecurringDialogTitle] = 'Edit Recurring Item',
    _a[exports.editorSave] = 'Save',
    _a[exports.editorCancel] = 'Cancel',
    _a[exports.editorDelete] = 'Delete',
    _a[exports.editorTitle] = 'Event',
    _a[exports.editorEventTitle] = 'Title',
    _a[exports.editorEventStart] = 'Start',
    _a[exports.editorEventStartTimeZone] = 'Start Time Zone',
    _a[exports.editorEventEnd] = 'End',
    _a[exports.editorEventEndTimeZone] = 'End Time Zone',
    _a[exports.editorEventAllDay] = 'All Day Event',
    _a[exports.editorEventDescription] = 'Description',
    _a[exports.editorEventSeparateTimeZones] = 'End in different Time Zone',
    _a[exports.editorEventTimeZone] = 'Specify Time Zone',
    _a[exports.recurrenceEditorRepeat] = 'Repeat',
    _a[exports.recurrenceEditorDailyInterval] = 'day(s)',
    _a[exports.recurrenceEditorDailyRepeatEvery] = 'Repeat every',
    _a[exports.recurrenceEditorWeeklyInterval] = 'week(s)',
    _a[exports.recurrenceEditorWeeklyRepeatEvery] = 'Repeat every',
    _a[exports.recurrenceEditorWeeklyRepeatOn] = 'Repeat on',
    _a[exports.recurrenceEditorMonthlyDay] = 'Day',
    _a[exports.recurrenceEditorMonthlyInterval] = 'month(s)',
    _a[exports.recurrenceEditorMonthlyRepeatEvery] = 'Repeat every',
    _a[exports.recurrenceEditorMonthlyRepeatOn] = 'Repeat on',
    _a[exports.recurrenceEditorYearlyOf] = 'of',
    _a[exports.recurrenceEditorYearlyRepeatEvery] = 'Repeat every',
    _a[exports.recurrenceEditorYearlyRepeatOn] = 'Repeat on',
    _a[exports.recurrenceEditorYearlyInterval] = 'year(s)',
    _a[exports.recurrenceEditorFrequenciesDaily] = 'Daily',
    _a[exports.recurrenceEditorFrequenciesMonthly] = 'Monthly',
    _a[exports.recurrenceEditorFrequenciesNever] = 'Never',
    _a[exports.recurrenceEditorFrequenciesWeekly] = 'Weekly',
    _a[exports.recurrenceEditorFrequenciesYearly] = 'Yearly',
    _a[exports.recurrenceEditorOffsetPositionsFirst] = 'First',
    _a[exports.recurrenceEditorOffsetPositionsSecond] = 'Second',
    _a[exports.recurrenceEditorOffsetPositionsThird] = 'Third',
    _a[exports.recurrenceEditorOffsetPositionsFourth] = 'Fourth',
    _a[exports.recurrenceEditorOffsetPositionsLast] = 'Last',
    _a[exports.recurrenceEditorWeekdaysDay] = 'Day',
    _a[exports.recurrenceEditorWeekdaysWeekday] = 'Weekday',
    _a[exports.recurrenceEditorWeekdaysWeekendday] = 'Weekend Day',
    _a[exports.recurrenceEditorEndAfter] = 'After',
    _a[exports.recurrenceEditorEndOccurrence] = 'occurrence(s)',
    _a[exports.recurrenceEditorEndLabel] = 'End',
    _a[exports.recurrenceEditorEndNever] = 'Never',
    _a[exports.recurrenceEditorEndOn] = 'On',
    _a[exports.deleteConfirmation] = 'Are you sure you want to delete this event?',
    _a[exports.deleteRecurringConfirmation] = 'Do you want to delete only this event occurrence or the whole series?',
    _a[exports.deleteOccurrence] = 'Delete current occurrence',
    _a[exports.deleteSeries] = 'Delete the series',
    _a[exports.deleteDialogTitle] = 'Delete Event',
    _a[exports.deleteRecurringDialogTitle] = 'Delete Recurring Item',
    _a[exports.dateTitle] = 'Date',
    _a[exports.timeTitle] = 'Time',
    _a[exports.eventTitle] = 'Event',
    _a[exports.noEvents] = 'no events',
    _a[exports.moreEvents] = 'More events',
    _a);
