// @progress/kendo-react-scheduler/dist/npm/services/slotsService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toSlots = void 0;
var React = require("react");
var kendo_date_math_1 = require("@progress/kendo-date-math");
var constants_1 = require("../constants");
/** @hidden */
var toSlots = function (_dateRange, _a, _b) {
    var step = _a.step;
    var groups = _b.groups, ranges = _b.ranges;
    var slots = [];
    groups.forEach(function (group) {
        ranges.forEach(function (range) {
            var viewStart = range.zonedStart;
            var viewEnd = range.zonedEnd;
            var offset = (viewEnd.timezoneOffset - viewStart.timezoneOffset);
            for (var current = viewStart.clone(), index = 0, deficiency = offset < 0 ? offset * constants_1.MS_PER_MINUTE * -1 : 0; current.getTime() < viewEnd.getTime(); index++, current = current.addTime(step)) {
                var _ref = React.createRef();
                var zonedStart = current.clone();
                var zonedEnd = current.addTime(step);
                var difference = (zonedEnd.timezoneOffset - zonedStart.timezoneOffset) * constants_1.MS_PER_MINUTE;
                var canFill = step <= Math.abs(difference);
                if (deficiency && canFill) {
                    deficiency -= step;
                    current = current.addTime(-step);
                }
                var start = new Date(zonedStart.getTime());
                var end = new Date(zonedEnd.getTime());
                var slot = {
                    _ref: _ref,
                    index: index,
                    end: end,
                    start: start,
                    zonedStart: zonedStart,
                    zonedEnd: zonedEnd,
                    range: range,
                    group: group,
                    items: [],
                    isAllDay: kendo_date_math_1.MS_PER_DAY <= step
                };
                slots.push(slot);
            }
        });
    });
    return slots;
};
exports.toSlots = toSlots;
