// @progress/kendo-react-scheduler/dist/npm/recurrence/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getUntilMin = exports.getUntilEnd = exports.getMonthDay = exports.getMonth = exports.getRepeatOnRule = exports.getEndRule = exports.getOffsets = exports.getOffset = exports.getExtendedWeekDays = exports.getWeekDay = exports.getWeekDays = exports.getMonths = exports.getUntil = exports.getFrequency = exports.getFrequencies = exports.getRule = exports.weekDayRuleFromString = exports.weekendRule = exports.weekdayRule = exports.dayRule = exports.FREQUENCIES = exports.OFFSET_POSITIONS = exports.capitalize = void 0;
var utils_1 = require("../utils");
var kendo_recurrence_1 = require("@progress/kendo-recurrence");
var kendo_date_math_1 = require("@progress/kendo-date-math");
/** @hidden */
var capitalize = function (value) { return value.charAt(0).toUpperCase() + value.slice(1); };
exports.capitalize = capitalize;
/** @hidden */
exports.OFFSET_POSITIONS = ['first', 'second', 'third', 'fourth', 'last'];
/** @hidden */
exports.FREQUENCIES = ['never', 'daily', 'weekly', 'monthly', 'yearly'];
/** @hidden */
exports.dayRule = [
    { day: 0, offset: 0 },
    { day: 1, offset: 0 },
    { day: 2, offset: 0 },
    { day: 3, offset: 0 },
    { day: 4, offset: 0 },
    { day: 5, offset: 0 },
    { day: 6, offset: 0 }
];
/** @hidden */
exports.weekdayRule = [
    { day: 1, offset: 0 },
    { day: 2, offset: 0 },
    { day: 3, offset: 0 },
    { day: 4, offset: 0 },
    { day: 5, offset: 0 }
];
/** @hidden */
exports.weekendRule = [
    { day: 0, offset: 0 },
    { day: 6, offset: 0 }
];
/** @hidden */
var weekDayRuleFromString = function (weekDay) {
    switch (weekDay) {
        case 'day':
            return exports.dayRule;
        case 'weekday':
            return exports.weekdayRule;
        case 'weekend':
            return exports.weekendRule;
        default:
            break;
    }
    return;
};
exports.weekDayRuleFromString = weekDayRuleFromString;
/** @hidden */
var getRule = function (recurrenceRule) {
    return (0, kendo_recurrence_1.parseRule)({ recurrenceRule: recurrenceRule }) || {};
};
exports.getRule = getRule;
/** @hidden */
var getFrequencies = function (localizeMessage) {
    return exports.FREQUENCIES.map(function (freq) { return ({
        value: freq,
        text: localizeMessage('Frequencies' + (0, exports.capitalize)(freq))
    }); });
};
exports.getFrequencies = getFrequencies;
/** @hidden */
var getFrequency = function (rrule) {
    if ((0, utils_1.isPresent)(rrule) && !(0, utils_1.isNullOrEmptyString)(rrule.freq)) {
        return rrule.freq;
    }
    return 'never';
};
exports.getFrequency = getFrequency;
/** @hidden */
var getUntil = function (rrule) {
    if ((0, utils_1.isPresent)(rrule.until)) {
        return (0, kendo_date_math_1.toLocalDate)(rrule.until);
    }
    return null;
};
exports.getUntil = getUntil;
/** @hidden */
var getMonths = function (intl) {
    return intl.dateFormatNames({
        type: 'months',
        nameType: 'wide',
        standAlone: true
    }).map(function (month, idx) { return ({
        text: month,
        value: idx + 1
    }); });
};
exports.getMonths = getMonths;
/** @hidden */
var getWeekDays = function (intl) {
    var firstDay = intl.firstDay();
    var abbrNames = intl.dateFormatNames({
        type: 'days',
        nameType: 'abbreviated'
    }).map(function (day, idx) { return ({
        text: day,
        value: idx
    }); });
    /* Sorting according to first week day */
    return (abbrNames.slice(firstDay)).concat(abbrNames.slice(0, firstDay));
};
exports.getWeekDays = getWeekDays;
/** @hidden */
var getWeekDay = function (weekDays, rrule, start) {
    return weekDays.find(function (item) {
        var value;
        if ((0, utils_1.isPresent)(rrule.byWeekDay)) {
            var weekDaysCount = rrule.byWeekDay.length;
            switch (weekDaysCount) {
                case 7:
                    value = 'day';
                    break;
                case 5:
                    value = 'weekday';
                    break;
                case 2:
                    value = 'weekend';
                    break;
                case 1:
                    value = rrule.byWeekDay[0].day;
                    break;
                default:
                    value = start.getDay();
                    break;
            }
        }
        return value === item.value;
    }) || weekDays[0];
};
exports.getWeekDay = getWeekDay;
/** @hidden */
var getExtendedWeekDays = function (intl, localizeMessage) {
    var firstDay = intl.firstDay();
    var wideNames = intl.dateFormatNames({
        type: 'days',
        nameType: 'wide'
    }).map(function (day, idx) { return ({
        text: day,
        value: idx
    }); });
    var sortedWideNames = (wideNames.slice(firstDay)).concat(wideNames.slice(0, firstDay));
    var specialRules = [
        { text: localizeMessage('WeekdaysDay'), value: 'day' },
        { text: localizeMessage('WeekdaysWeekday'), value: 'weekday' },
        { text: localizeMessage('WeekdaysWeekendday'), value: 'weekend' }
    ];
    return specialRules.concat(sortedWideNames);
};
exports.getExtendedWeekDays = getExtendedWeekDays;
/** @hidden */
var getOffset = function (offsets, rrule) {
    return offsets.find(function (offset) { return offset.value === (rrule.bySetPosition
        ? rrule.bySetPosition[0]
        : rrule.byWeekDay && rrule.byWeekDay[0]
            ? rrule.byWeekDay[0].offset
            : 1); }) || offsets[0];
};
exports.getOffset = getOffset;
/** @hidden */
var getOffsets = function (localizeMessage) {
    var values = [1, 2, 3, 4, -1];
    return exports.OFFSET_POSITIONS.map(function (offset, idx) { return ({
        text: localizeMessage('OffsetPositions' + (0, exports.capitalize)(offset)),
        value: values[idx]
    }); });
};
exports.getOffsets = getOffsets;
/** @hidden */
var getEndRule = function (rrule) {
    if ((0, utils_1.isPresent)(rrule.count)) {
        return 'count';
    }
    else if ((0, utils_1.isPresent)(rrule.until)) {
        return 'until';
    }
    else {
        return 'never';
    }
};
exports.getEndRule = getEndRule;
/** @hidden */
var getRepeatOnRule = function (rrule) {
    if ((0, utils_1.isPresent)(rrule.byWeekDay)) {
        return 'weekday';
    }
    else if ((0, utils_1.isPresent)(rrule.byMonthDay)) {
        return 'monthday';
    }
};
exports.getRepeatOnRule = getRepeatOnRule;
/** @hidden */
var getMonth = function (months, rrule) {
    return months.find(function (month) { return (rrule.byMonth ? rrule.byMonth[0] : null) === month.value; }) || months[0];
};
exports.getMonth = getMonth;
/** @hidden */
var getMonthDay = function (rrule, start) {
    if ((0, utils_1.isPresent)(rrule.byMonthDay) && rrule.byMonthDay.length > 0) {
        return rrule.byMonthDay[0];
    }
    else {
        return start.getDate();
    }
};
exports.getMonthDay = getMonthDay;
/** @hidden */
var getUntilEnd = function (start, until) {
    var currentStart = start;
    var currentUntil = until;
    if ((0, utils_1.isPresent)(currentUntil)) {
        return currentUntil;
    }
    else {
        return new Date(currentStart.getFullYear(), currentStart.getMonth(), currentStart.getDate(), 23, 59, 59);
    }
};
exports.getUntilEnd = getUntilEnd;
/** @hidden */
var getUntilMin = function (start, until) {
    var currentStart = start;
    var currentUntil = until;
    if ((0, utils_1.isPresent)(currentUntil) && currentUntil < currentStart) {
        return currentUntil;
    }
    else {
        return currentStart;
    }
};
exports.getUntilMin = getUntilMin;
