// @progress/kendo-react-scheduler/dist/npm/views/time/TimelineView.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultProps = exports.TimelineView = void 0;
var React = require("react");
var MultiDayTimelineView_1 = require("./MultiDayTimelineView");
var messages_1 = require("../../messages");
/**
 * Represents the KendoReact [TimelineView]({% slug timeline_view_scheduler %}) Component.
 */
var TimelineView = function (props) { return React.createElement(MultiDayTimelineView_1.MultiDayTimelineView, __assign({}, props)); };
exports.TimelineView = TimelineView;
/** @hidden */
exports.defaultProps = {
    name: 'timeline',
    title: function (localization) { return localization.toLanguageString(messages_1.timelineViewTitle, messages_1.messages[messages_1.timelineViewTitle]); },
    dateRange: MultiDayTimelineView_1.defaultProps.dateRange,
    step: 1,
    slotDuration: 60,
    slotDivisions: 2,
    numberOfDays: 1,
    selectedDateFormat: '{0:D}',
    selectedShortDateFormat: '{0:d}'
};
exports.TimelineView.defaultProps = exports.defaultProps;
exports.TimelineView.displayName = 'KendoReactSchedulerTimelineView';
