// @progress/kendo-react-scheduler/dist/npm/Scheduler.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Scheduler = exports.DATA_ACTION = void 0;
var React = require("react");
var PropTypes = require("prop-types");
var utils_1 = require("./utils");
var kendo_react_common_1 = require("@progress/kendo-react-common");
var SchedulerContext_1 = require("./context/SchedulerContext");
var useControlledState_1 = require("./hooks/useControlledState");
var kendo_react_intl_1 = require("@progress/kendo-react-intl");
var SchedulerHeader_1 = require("./components/header/SchedulerHeader");
var SchedulerNavigation_1 = require("./components/header/navigation/SchedulerNavigation");
var SchedulerViewSelector_1 = require("./components/header/view-selector/SchedulerViewSelector");
var kendo_date_math_1 = require("@progress/kendo-date-math");
var SchedulerFooter_1 = require("./components/footer/SchedulerFooter");
var kendo_react_buttons_1 = require("@progress/kendo-react-buttons");
var messages_1 = require("./messages");
var NavigationDatePicker_1 = require("./components/header/navigation/NavigationDatePicker");
var ViewSelectorList_1 = require("./components/header/view-selector/ViewSelectorList");
var BussinessHours_1 = require("./components/footer/bussiness-hours/BussinessHours");
var main_1 = require("./main");
var kendo_react_common_2 = require("@progress/kendo-react-common");
var package_metadata_1 = require("./package-metadata");
var DEFAULT_DATE_FORMAT = '{0:D}';
var DEFAULT_SHORT_DATE_FORMAT = '{0:d}';
/**
 * Specifies the available [DataAction]({% slug api_scheduler_data_action %}) types.
 */
var DATA_ACTION;
(function (DATA_ACTION) {
    DATA_ACTION[DATA_ACTION["create"] = 0] = "create";
    DATA_ACTION[DATA_ACTION["update"] = 1] = "update";
    DATA_ACTION[DATA_ACTION["remove"] = 2] = "remove";
})(DATA_ACTION = exports.DATA_ACTION || (exports.DATA_ACTION = {}));
// tslint:enable:max-line-length
/**
 * Represents the [KendoReact Scheduler component]({% slug overview_scheduler %})
 */
exports.Scheduler = React.forwardRef(function (props, ref) {
    (0, kendo_react_common_2.validatePackage)(package_metadata_1.packageMetadata);
    var timezone = props.timezone, onDataChange = props.onDataChange;
    var element = React.useRef(null);
    var scheduler = React.useRef(null);
    React.useImperativeHandle(scheduler, function () { return ({ props: props, element: element.current }); });
    React.useImperativeHandle(ref, function () { return scheduler.current; });
    var dir = (0, kendo_react_common_1.useRtl)(element);
    var intl = (0, kendo_react_intl_1.useInternationalization)();
    var localization = (0, kendo_react_intl_1.useLocalization)();
    var fields = React.useMemo(function () { return (0, utils_1.getModelFields)(props.modelFields); }, [props.modelFields]).fields;
    var _a = (0, useControlledState_1.useControlledState)(props.defaultDate || defaultProps.defaultDate, props.date, props.onDateChange), date = _a[0], setDate = _a[1];
    var views = React.Children.toArray(props.children || []);
    var _b = (0, useControlledState_1.useControlledState)(props.defaultView || (views[0] && views[0].props.name) || 'day', props.view, props.onViewChange), activeViewName = _b[0], setActiveViewName = _b[1];
    var _c = (0, useControlledState_1.useControlledState)(true), showWorkHours = _c[0], setShowWorkHours = _c[1];
    var view = views.find(function (currentView) { return currentView.props.name === activeViewName; })
        || views[0]
        || React.createElement(main_1.DayView, null);
    var data = props.data || defaultProps.data;
    var groups = (0, utils_1.toSchedulerGroups)(props.group, props.resources);
    var orientation = props.group && props.group.orientation ? props.group.orientation : 'horizontal';
    var dateFormat = view.props.selectedDateFormat || DEFAULT_DATE_FORMAT;
    var shortDateFormat = view.props.selectedShortDateFormat || DEFAULT_SHORT_DATE_FORMAT;
    var slotDuration = view.props.slotDuration;
    var dateRange = (view.props.dateRange !== undefined
        ? typeof view.props.dateRange === 'function'
            ? view.props.dateRange.call(undefined, {
                intl: intl,
                date: date,
                timezone: timezone,
                numberOfDays: view.props.numberOfDays,
                workWeekStart: view.props.workWeekStart || intl.firstDay(),
                workWeekEnd: view.props.workWeekEnd || (intl.firstDay() + view.props.numberOfDays) % 6
            })
            : view.props.dateRange
        : { start: defaultProps.defaultDate, end: defaultProps.defaultDate });
    var handleDataChange = React.useCallback(function (_a) {
        var _b = _a.created, created = _b === void 0 ? [] : _b, _c = _a.updated, updated = _c === void 0 ? [] : _c, _d = _a.deleted, deleted = _d === void 0 ? [] : _d;
        if (onDataChange) {
            var args = {
                created: created,
                updated: updated,
                deleted: deleted
            };
            onDataChange.call(undefined, args);
        }
    }, [onDataChange]);
    var handleCreate = React.useCallback(function (action) {
        var created = [action.dataItem];
        handleDataChange({ created: created });
    }, [handleDataChange]);
    var handleUpdate = React.useCallback(function (action) {
        var created = [];
        var updated = [];
        if (action.series) {
            if (Array.isArray(action.dataItem)) {
                action.dataItem.map(function (dataItem) {
                    var masterClone = (0, kendo_react_common_1.clone)((0, utils_1.findMaster)(action.dataItem, fields, data));
                    var newDataItem = (0, kendo_react_common_1.clone)(dataItem);
                    (0, utils_1.setField)(newDataItem, fields.originalStart, (0, utils_1.getField)(masterClone, fields.originalStart));
                    (0, utils_1.setField)(newDataItem, fields.recurrenceId, (0, utils_1.getField)(masterClone, fields.recurrenceId));
                    (0, utils_1.setField)(newDataItem, fields.recurrenceExceptions, (0, utils_1.getField)(masterClone, fields.recurrenceExceptions));
                    updated.push(newDataItem);
                });
            }
            else {
                var masterClone = (0, kendo_react_common_1.clone)((0, utils_1.findMaster)(action.dataItem, fields, data));
                var newDataItem = (0, kendo_react_common_1.clone)(action.dataItem);
                (0, utils_1.setField)(newDataItem, fields.originalStart, (0, utils_1.getField)(masterClone, fields.originalStart));
                (0, utils_1.setField)(newDataItem, fields.recurrenceId, (0, utils_1.getField)(masterClone, fields.recurrenceId));
                (0, utils_1.setField)(newDataItem, fields.recurrenceExceptions, (0, utils_1.getField)(masterClone, fields.recurrenceExceptions));
                updated.push(newDataItem);
            }
        }
        else {
            if (Array.isArray(action.dataItem)) {
                action.dataItem.map(function (dataItem) {
                    var isException = (0, utils_1.getField)(dataItem, fields.recurrenceRule) !== null
                        && (0, utils_1.getField)(dataItem, fields.recurrenceRule) !== undefined;
                    var isRecurring = (0, utils_1.getField)(dataItem, fields.recurrenceId) !== null
                        && (0, utils_1.getField)(dataItem, fields.recurrenceId) !== undefined;
                    if (isRecurring && isException) {
                        var masterClone = (0, kendo_react_common_1.clone)((0, utils_1.findMaster)(dataItem, fields, data));
                        var exceptionDate = (0, utils_1.getField)(dataItem, fields.originalStart);
                        var currentExceptions = (0, utils_1.getField)(masterClone, fields.recurrenceExceptions) || [];
                        (0, utils_1.setField)(masterClone, fields.recurrenceExceptions, __spreadArray(__spreadArray([], currentExceptions, true), [exceptionDate], false));
                        (0, utils_1.setField)(dataItem, fields.recurrenceRule, null);
                        updated.push(masterClone);
                        created.push(dataItem);
                    }
                    else {
                        updated.push(dataItem);
                    }
                });
            }
            else {
                var isException = (0, utils_1.getField)(action.dataItem, fields.recurrenceRule) !== null
                    && (0, utils_1.getField)(action.dataItem, fields.recurrenceRule) !== undefined;
                var isRecurring = (0, utils_1.getField)(action.dataItem, fields.recurrenceId) !== null
                    && (0, utils_1.getField)(action.dataItem, fields.recurrenceId) !== undefined;
                if (isRecurring && isException) {
                    var masterClone = (0, kendo_react_common_1.clone)((0, utils_1.findMaster)(action.dataItem, fields, data));
                    var exceptionDate = (0, utils_1.getField)(action.dataItem, fields.originalStart);
                    var currentExceptions = (0, utils_1.getField)(masterClone, fields.recurrenceExceptions) || [];
                    (0, utils_1.setField)(masterClone, fields.recurrenceExceptions, __spreadArray(__spreadArray([], currentExceptions, true), [exceptionDate], false));
                    (0, utils_1.setField)(action.dataItem, fields.recurrenceRule, null);
                    updated.push(masterClone);
                    created.push(action.dataItem);
                }
                else {
                    updated.push(action.dataItem);
                }
            }
        }
        handleDataChange({ updated: updated, created: created });
    }, [handleDataChange, fields, data]);
    var handleRemove = React.useCallback(function (action) {
        var updated = [];
        var deleted = [];
        if (action.series) {
            var masterClone = (0, kendo_react_common_1.clone)((0, utils_1.findMaster)(action.dataItem, fields, data));
            var dataItem = (0, kendo_react_common_1.clone)(action.dataItem);
            (0, utils_1.setField)(dataItem, fields.originalStart, (0, utils_1.getField)(masterClone, fields.originalStart));
            (0, utils_1.setField)(dataItem, fields.recurrenceId, (0, utils_1.getField)(masterClone, fields.recurrenceId));
            (0, utils_1.setField)(dataItem, fields.recurrenceRule, (0, utils_1.getField)(masterClone, fields.recurrenceRule));
            (0, utils_1.setField)(dataItem, fields.recurrenceExceptions, (0, utils_1.getField)(masterClone, fields.recurrenceExceptions));
            deleted.push(dataItem);
        }
        else {
            var isException = (0, utils_1.getField)(action.dataItem, fields.recurrenceRule) !== null
                && (0, utils_1.getField)(action.dataItem, fields.recurrenceRule) !== undefined;
            if (!isException) {
                deleted.push(action.dataItem);
            }
            else {
                var masterClone = (0, kendo_react_common_1.clone)((0, utils_1.findMaster)(action.dataItem, fields, data));
                var exceptionDate = (0, utils_1.getField)(action.dataItem, fields.originalStart);
                var currentExceptions = (0, utils_1.getField)(masterClone, fields.recurrenceExceptions) || [];
                (0, utils_1.setField)(masterClone, fields.recurrenceExceptions, __spreadArray(__spreadArray([], currentExceptions, true), [exceptionDate], false));
                (0, utils_1.setField)(action.dataItem, fields.recurrenceRule, null);
                updated.push(masterClone);
            }
        }
        handleDataChange({ updated: updated, deleted: deleted });
    }, [handleDataChange, fields, data]);
    var handleDataAction = React.useCallback(function (action) {
        switch (action.type) {
            case DATA_ACTION.create:
                handleCreate(action);
                break;
            case DATA_ACTION.update:
                handleUpdate(action);
                break;
            case DATA_ACTION.remove:
                handleRemove(action);
                break;
            default:
                break;
        }
    }, [handleCreate, handleRemove, handleUpdate]);
    var handleActiveViewNameChange = React.useCallback(function (newView, event) {
        setActiveViewName(newView, __assign(__assign({}, event), { target: scheduler.current }));
    }, [
        setActiveViewName,
        scheduler
    ]);
    var handleDateChange = React.useCallback(function (newDate, event) {
        setDate(newDate, __assign(__assign({}, event), { target: scheduler.current }));
    }, [
        setDate,
        scheduler
    ]);
    var handleDatePickerChange = React.useCallback(function (event) {
        if (!event.value) {
            return;
        }
        setDate(event.value, __assign(__assign({}, event), { target: scheduler.current, nativeEvent: event.nativeEvent }));
    }, [
        setDate,
        scheduler
    ]);
    var handleNextClick = React.useCallback(function (syntheticEvent) {
        syntheticEvent.preventDefault();
        var offset = view.props.numberOfDays || 1;
        var isMonthView = offset > 27;
        var newDate = isMonthView
            ? (0, kendo_date_math_1.addMonths)(date, Math.round(offset / 27))
            : (0, kendo_date_math_1.addDays)(date, offset);
        setDate(newDate, event);
    }, [date, setDate, view.props.numberOfDays]);
    var handlePrevClick = React.useCallback(function (syntheticEvent) {
        syntheticEvent.preventDefault();
        var offset = view.props.numberOfDays || 1;
        var isMonthView = offset > 27;
        var newDate = isMonthView
            ? (0, kendo_date_math_1.addMonths)(date, -(Math.round(offset / 27)))
            : (0, kendo_date_math_1.addDays)(date, -(offset));
        setDate(newDate, event);
    }, [date, setDate, view.props.numberOfDays]);
    var handleTodayClick = React.useCallback(function (syntheticEvent) {
        syntheticEvent.preventDefault();
        var newDate = (0, utils_1.getToday)();
        setDate(newDate, event);
    }, [setDate]);
    var handleShowWorkHoursClick = React.useCallback(function () { setShowWorkHours(!showWorkHours); }, [setShowWorkHours, showWorkHours]);
    var handleFocus = React.useCallback(function () {
        if (element.current) {
            element.current.style.boxShadow = '0 0.5px 0.5px 0.5px rgba(0, 0, 0, .12)';
        }
    }, [element]);
    var handleBlur = React.useCallback(function () {
        if (element.current) {
            element.current.style.boxShadow = '';
        }
    }, [element]);
    var style = React.useMemo(function () { return (__assign(__assign({}, props.style), { height: props.height })); }, [props.height, props.style]);
    var className = React.useMemo(function () { return (0, kendo_react_common_1.classNames)({ 'k-rtl': props.rtl !== undefined ? props.rtl : dir === 'rtl' }, 'k-widget k-scheduler k-floatwrap', props.className); }, [props.className, props.rtl, dir]);
    var todayText = localization.toLanguageString(messages_1.today, messages_1.messages[messages_1.today]);
    var previousText = localization.toLanguageString(messages_1.previousTitle, messages_1.messages[messages_1.previousTitle]);
    var nextText = localization.toLanguageString(messages_1.nextTitle, messages_1.messages[messages_1.nextTitle]);
    var Header = view.props.header || props.header || defaultProps.header;
    var Footer = view.props.footer || props.footer || defaultProps.footer;
    var Navigation = SchedulerNavigation_1.SchedulerNavigation;
    var ViewSelector = SchedulerViewSelector_1.SchedulerViewSelector;
    var _d = React.useState(null), eventSelection = _d[0], setEventSelection = _d[1];
    return (React.createElement(SchedulerContext_1.SchedulerContext
    // Static
    , { 
        // Static
        element: element, props: props, views: views, fields: fields, groups: groups, dateRange: dateRange, orientation: orientation, dateFormat: { dateFormat: dateFormat, shortDateFormat: shortDateFormat }, 
        // State
        date: [date, handleDateChange], activeView: [activeViewName, handleActiveViewNameChange], selection: [eventSelection, setEventSelection], 
        // Reducers
        data: [data, handleDataAction] },
        React.createElement("div", { ref: element, id: props.id, style: style, className: className, tabIndex: props.tabIndex, 
            // Aria
            dir: dir, role: props.role, "aria-label": props.ariaLabel, "aria-labelledby": props.ariaLabelledby, "aria-activedescendant": (eventSelection && eventSelection.props.id) || undefined, 
            // Handlers
            onFocus: handleFocus, onBlur: handleBlur },
            React.createElement(Header, null,
                React.createElement(Navigation, null,
                    React.createElement(kendo_react_buttons_1.ButtonGroup, { className: "k-scheduler-navigation" },
                        React.createElement(kendo_react_buttons_1.Button, { role: "button", tabIndex: -1, title: todayText, "aria-label": todayText, onClick: handleTodayClick }, todayText),
                        React.createElement(kendo_react_buttons_1.Button, { role: "button", tabIndex: -1, icon: 'caret-alt-left', title: previousText, "aria-label": previousText, onClick: handlePrevClick }),
                        React.createElement(kendo_react_buttons_1.Button, { role: "button", tabIndex: -1, icon: 'caret-alt-right', title: nextText, "aria-label": nextText, onClick: handleNextClick }))),
                React.createElement(NavigationDatePicker_1.NavigationDatePicker, { value: date, onChange: handleDatePickerChange }),
                React.createElement(kendo_react_buttons_1.ToolbarSpacer, null),
                React.createElement(ViewSelector, null,
                    React.createElement(ViewSelectorList_1.ViewSelectorList, null))),
            view && (React.createElement(view.type, __assign({ editable: props.editable, key: view.props.name, item: props.item, viewItem: props.viewItem, editItem: props.editItem, task: props.task, viewTask: props.viewTask, editTask: props.viewTask, slot: props.slot, viewSlot: props.viewSlot, editSlot: props.editSlot, form: props.form, onDataAction: handleDataAction, showWorkHours: showWorkHours }, view.props))),
            React.createElement(Footer, null, (slotDuration && slotDuration < (24 * 60)) && (React.createElement(BussinessHours_1.BusinessHours, null,
                React.createElement(kendo_react_buttons_1.Button, { tabIndex: -1, onClick: handleShowWorkHoursClick, icon: "clock" }, localization.toLanguageString(showWorkHours
                    ? messages_1.showFullDay
                    : messages_1.showWorkDay, messages_1.messages[showWorkHours
                    ? messages_1.showFullDay
                    : messages_1.showWorkDay]))))))));
});
var defaultProps = {
    data: [],
    height: 600,
    tabIndex: -1,
    editable: false,
    defaultDate: new Date(),
    header: SchedulerHeader_1.SchedulerHeader,
    footer: SchedulerFooter_1.SchedulerFooter,
    navigation: SchedulerNavigation_1.SchedulerNavigation,
    viewSelector: SchedulerViewSelector_1.SchedulerViewSelector
};
exports.Scheduler.propTypes = {
    data: PropTypes.array,
    editable: PropTypes.oneOfType([
        PropTypes.bool,
        PropTypes.shape({
            add: PropTypes.bool,
            drag: PropTypes.bool,
            edit: PropTypes.bool,
            remove: PropTypes.bool,
            resize: PropTypes.bool
        })
    ]),
    view: PropTypes.string,
    defaultView: PropTypes.string,
    date: PropTypes.any,
    defaultDate: PropTypes.any,
    rtl: PropTypes.bool,
    height: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    children: PropTypes.any,
    timezone: PropTypes.string,
    group: PropTypes.any,
    resources: PropTypes.any,
    modelFields: PropTypes.shape({
        id: PropTypes.string,
        start: PropTypes.string,
        startTimezone: PropTypes.string,
        end: PropTypes.string,
        endTimezone: PropTypes.string,
        isAllDay: PropTypes.string,
        title: PropTypes.string,
        description: PropTypes.string,
        recurrenceRule: PropTypes.string,
        recurrenceId: PropTypes.string,
        recurrenceException: PropTypes.string
    })
};
exports.Scheduler.defaultProps = defaultProps;
exports.Scheduler.displayName = 'KendoReactScheduler';
