// @progress/kendo-react-scheduler/dist/npm/items/hooks/use-remove-item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useRemoveItem = exports.REMOVE_ITEM_ACTION = void 0;
var useControlledState_1 = require("../../hooks/useControlledState");
var Scheduler_1 = require("../../Scheduler");
/**
 * Represents the available `action` types for the `SchedulerItemRemoveItemAction` objects.
 *
 * The available types are:
 * - `REMOVE_ITEM_SET`
 * - `REMOVE_ITEM_RESET`
 * - `REMOVE_ITEM_COMPLETE`
 */
var REMOVE_ITEM_ACTION;
(function (REMOVE_ITEM_ACTION) {
    REMOVE_ITEM_ACTION["set"] = "REMOVE_ITEM_SET";
    REMOVE_ITEM_ACTION["reset"] = "REMOVE_ITEM_RESET";
    REMOVE_ITEM_ACTION["complete"] = "REMOVE_ITEM_COMPLETE";
})(REMOVE_ITEM_ACTION = exports.REMOVE_ITEM_ACTION || (exports.REMOVE_ITEM_ACTION = {}));
/** @hidden */
var useRemoveItem = function (config, state) {
    var _a = useControlledState_1.useControlledState.apply(void 0, state), removeItem = _a[0], setRemoveItem = _a[1];
    var handleRemoveItemAction = function (action) {
        var newRemoveItem;
        switch (action.type) {
            case REMOVE_ITEM_ACTION.set:
                newRemoveItem = action.payload;
                break;
            case REMOVE_ITEM_ACTION.reset:
                newRemoveItem = null;
                break;
            case REMOVE_ITEM_ACTION.complete:
                if (config.onDataAction && removeItem) {
                    config.onDataAction.call(undefined, {
                        type: Scheduler_1.DATA_ACTION.remove,
                        series: config.series,
                        dataItem: removeItem
                    });
                }
                newRemoveItem = null;
                break;
            default:
                newRemoveItem = removeItem;
                break;
        }
        setRemoveItem(newRemoveItem);
    };
    return [removeItem, setRemoveItem, handleRemoveItemAction];
};
exports.useRemoveItem = useRemoveItem;
