// @progress/kendo-react-scheduler/dist/npm/items/hooks/use-resize-item.js
"use strict";
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useResizeItem = exports.RESIZE_ITEM_ACTION = void 0;
var kendo_react_common_1 = require("@progress/kendo-react-common");
var React = require("react");
var context_1 = require("../../context");
var useControlledState_1 = require("../../hooks/useControlledState");
var Scheduler_1 = require("../../Scheduler");
var utils_1 = require("../../utils");
/**
 * Represents the available `action` types for the `SchedulerItemResizeItemAction` objects.
 *
 * The available types are:
 * - `RESIZE_ITEM_SET`
 * - `RESIZE_ITEM_START`
 * - `RESIZE_ITEM_START_DRAG`
 * - `RESIZE_ITEM_START_DRAG_SELECTED`
 * - `RESIZE_ITEM_END_DRAG`
 * - `RESIZE_ITEM_END_DRAG_SELECTED`
 * - `RESIZE_ITEM_COMPLETE`
 * - `RESIZE_ITEM_COMPLETE_OCCURRENCE`
 * - `RESIZE_ITEM_COMPLETE_SERIES`
 * - `RESIZE_ITEM_RESET`
 */
var RESIZE_ITEM_ACTION;
(function (RESIZE_ITEM_ACTION) {
    RESIZE_ITEM_ACTION["set"] = "RESIZE_ITEM_SET";
    RESIZE_ITEM_ACTION["start"] = "RESIZE_ITEM_START";
    RESIZE_ITEM_ACTION["startDrag"] = "RESIZE_ITEM_START_DRAG";
    RESIZE_ITEM_ACTION["startDragSelected"] = "RESIZE_ITEM_START_DRAG_SELECTED";
    RESIZE_ITEM_ACTION["endDrag"] = "RESIZE_ITEM_END_DRAG";
    RESIZE_ITEM_ACTION["endDragSelected"] = "RESIZE_ITEM_END_DRAG_SELECTED";
    RESIZE_ITEM_ACTION["complete"] = "RESIZE_ITEM_COMPLETE";
    RESIZE_ITEM_ACTION["completeOccurrence"] = "RESIZE_ITEM_COMPLETE_OCCURRENCE";
    RESIZE_ITEM_ACTION["completeSeries"] = "RESIZE_ITEM_COMPLETE_SERIES";
    RESIZE_ITEM_ACTION["reset"] = "RESIZE_ITEM_RESET";
})(RESIZE_ITEM_ACTION = exports.RESIZE_ITEM_ACTION || (exports.RESIZE_ITEM_ACTION = {}));
;
/** @hidden */
var useResizeItem = function (config, state) {
    var oldSlot = React.useRef(null);
    var _a = useControlledState_1.useControlledState.apply(void 0, state), resizeItem = _a[0], setResizeItem = _a[1];
    var selectedItems = (0, context_1.useSchedulerViewSelectedItemsContext)()[0];
    var fields = (0, context_1.useSchedulerFieldsContext)();
    var handleDragItemAction = function (action, event) {
        var newResizeItem = resizeItem;
        switch (action.type) {
            case RESIZE_ITEM_ACTION.set:
                newResizeItem = action.payload;
                break;
            case RESIZE_ITEM_ACTION.reset:
                newResizeItem = null;
                break;
            case RESIZE_ITEM_ACTION.start: {
                event.stopPropagation();
                var clientX = action.payload.x;
                var clientY = action.payload.y;
                var slot = (0, utils_1.slotDive)(clientX, clientY, 7);
                if (!slot) {
                    return;
                }
                if (slot === oldSlot.current) {
                    return;
                }
                var slotStart = slot.getAttribute('data-slot-start');
                var newStart = new Date(Number(slotStart));
                var dataItem = (0, kendo_react_common_1.clone)(config.dataItem);
                if (newStart >= (0, utils_1.getField)(dataItem, fields.end)) {
                    return;
                }
                (0, utils_1.setField)(dataItem, fields.start, newStart);
                oldSlot.current = slot;
                newResizeItem = dataItem;
                break;
            }
            case RESIZE_ITEM_ACTION.startDragSelected: {
                event.stopPropagation();
                var clientX = action.payload.x;
                var clientY = action.payload.y;
                var dataItem = (0, kendo_react_common_1.clone)(config.dataItem);
                var slot = (0, utils_1.slotDive)(clientX, clientY, 7);
                if (!slot) {
                    return;
                }
                if (slot === oldSlot.current) {
                    return;
                }
                var slotStart = slot.getAttribute('data-slot-start');
                var itemStart = new Date((0, utils_1.getField)(dataItem, fields.start));
                var newStart = new Date(Number(slotStart));
                var distance_1 = newStart.getTime() - itemStart.getTime();
                var unreachable_1 = false;
                var newResizeItems = selectedItems.map(function (si) {
                    if (!si.current) {
                        return null;
                    }
                    var selectedDataItem = (0, kendo_react_common_1.clone)(si.current.props.dataItem);
                    var selectedStart = new Date(si.current.props.start.getTime() + distance_1);
                    if (selectedStart >= (0, utils_1.getField)(selectedDataItem, fields.end)) {
                        unreachable_1 = true;
                        return;
                    }
                    (0, utils_1.setField)(selectedDataItem, fields.start, selectedStart);
                    return selectedDataItem;
                }).filter(Boolean);
                if (!unreachable_1) {
                    newResizeItem = __spreadArray([], newResizeItems, true);
                }
                break;
            }
            case RESIZE_ITEM_ACTION.startDrag: {
                var clientX = action.payload.x;
                var clientY = action.payload.y;
                var slot = (0, utils_1.slotDive)(clientX, clientY, 7);
                if (!slot) {
                    return;
                }
                if (slot === oldSlot.current) {
                    return;
                }
                var slotStart = slot.getAttribute('data-slot-start');
                var newStart = new Date(Number(slotStart));
                var dataItem = (0, kendo_react_common_1.clone)(config.dataItem);
                if (newStart >= dataItem.end) {
                    return;
                }
                (0, utils_1.setField)(dataItem, fields.start, newStart);
                oldSlot.current = slot;
                newResizeItem = dataItem;
                break;
            }
            case RESIZE_ITEM_ACTION.endDrag: {
                var clientX = action.payload.x;
                var clientY = action.payload.y;
                var slot = (0, utils_1.slotDive)(clientX, clientY, 7);
                if (!slot) {
                    return;
                }
                if (slot === oldSlot.current) {
                    return;
                }
                var slotEnd = slot.getAttribute('data-slot-end');
                var newEnd = new Date(Number(slotEnd));
                var dataItem = (0, kendo_react_common_1.clone)(config.dataItem);
                if (newEnd <= dataItem.start) {
                    return;
                }
                (0, utils_1.setField)(dataItem, fields.end, newEnd);
                oldSlot.current = slot;
                newResizeItem = dataItem;
                break;
            }
            case RESIZE_ITEM_ACTION.endDragSelected: {
                event.stopPropagation();
                var clientX = action.payload.x;
                var clientY = action.payload.y;
                var dataItem = (0, kendo_react_common_1.clone)(config.dataItem);
                var slot = (0, utils_1.slotDive)(clientX, clientY, 7);
                if (!slot) {
                    return;
                }
                if (slot === oldSlot.current) {
                    return;
                }
                var slotEnd = slot.getAttribute('data-slot-start');
                var itemEnd = new Date((0, utils_1.getField)(dataItem, fields.end));
                var newEnd = new Date(Number(slotEnd));
                var distance_2 = newEnd.getTime() - itemEnd.getTime();
                var unreachable_2;
                var newResizeItems = selectedItems.map(function (si) {
                    if (!si.current) {
                        return null;
                    }
                    var selectedDataItem = (0, kendo_react_common_1.clone)(si.current.props.dataItem);
                    var selectedEnd = new Date(si.current.props.end.getTime() + distance_2);
                    if (selectedEnd <= (0, utils_1.getField)(selectedDataItem, fields.start)) {
                        unreachable_2 = true;
                        return;
                    }
                    (0, utils_1.setField)(selectedDataItem, fields.end, selectedEnd);
                    return selectedDataItem;
                }).filter(Boolean);
                if (!unreachable_2) {
                    newResizeItem = __spreadArray([], newResizeItems, true);
                }
                break;
            }
            case RESIZE_ITEM_ACTION.complete: {
                newResizeItem = null;
                if (config.onDataAction && resizeItem) {
                    config.onDataAction.call(undefined, {
                        type: Scheduler_1.DATA_ACTION.update,
                        series: false,
                        dataItem: resizeItem
                    });
                }
                break;
            }
            case RESIZE_ITEM_ACTION.completeOccurrence: {
                newResizeItem = null;
                if (config.onDataAction && resizeItem) {
                    config.onDataAction.call(undefined, {
                        type: Scheduler_1.DATA_ACTION.update,
                        series: false,
                        dataItem: resizeItem
                    });
                }
                break;
            }
            case RESIZE_ITEM_ACTION.completeSeries: {
                newResizeItem = null;
                var newDataItem = void 0;
                if (Array.isArray(resizeItem)) {
                    newDataItem = resizeItem.map(function (item) {
                        var newItem = (0, kendo_react_common_1.clone)(item);
                        (0, utils_1.setField)(newItem, fields.start, (0, utils_1.getField)(item, fields.start));
                        (0, utils_1.setField)(newItem, fields.end, (0, utils_1.getField)(item, fields.end));
                        return newItem;
                    });
                }
                else {
                    var newItem = (0, kendo_react_common_1.clone)(resizeItem);
                    (0, utils_1.setField)(newItem, fields.start, (0, utils_1.getField)(resizeItem, fields.start));
                    (0, utils_1.setField)(newItem, fields.end, (0, utils_1.getField)(resizeItem, fields.end));
                    newDataItem = newItem;
                }
                if (config.onDataAction && newDataItem) {
                    config.onDataAction.call(undefined, {
                        type: Scheduler_1.DATA_ACTION.update,
                        series: true,
                        dataItem: newDataItem
                    });
                }
                break;
            }
            default:
                newResizeItem = null;
                break;
        }
        setResizeItem(newResizeItem);
    };
    return [resizeItem, setResizeItem, handleDragItemAction];
};
exports.useResizeItem = useResizeItem;
