// @progress/kendo-react-scheduler/dist/npm/constants/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_RANGE = exports.MAX_DATE = exports.MIN_DATE = exports.DEFAULT_GROUP = exports.VISIBILITY_HIDDEN_STYLES = exports.DISPLAY_NONE_STYLES = exports.DEFAULT_ID = exports.OCCURRENCE_ID = exports.EMPTY_RESOURCE = exports.HOURS_IN_DAY_COUNT = exports.WEEKS_COUNT = exports.DAYS_IN_WEEK_COUNT = exports.DEFAULT_EVENT_HEIGHT = exports.BORDER_WIDTH = exports.TIME_FORMAT = exports.MS_PER_HOUR = exports.MS_PER_MINUTE = exports.MS_PER_SECOND = exports.MIDNIGHT_INVARIANT = exports.MS_PER_DAY = void 0;
var kendo_date_math_1 = require("@progress/kendo-date-math");
/** @hidden */
var kendo_date_math_2 = require("@progress/kendo-date-math");
Object.defineProperty(exports, "MS_PER_DAY", { enumerable: true, get: function () { return kendo_date_math_2.MS_PER_DAY; } });
/** @hidden */
exports.MIDNIGHT_INVARIANT = new Date(1980, 0, 1);
/** @hidden */
exports.MS_PER_SECOND = 1000;
/** @hidden */
exports.MS_PER_MINUTE = 60 * exports.MS_PER_SECOND;
/** @hidden */
exports.MS_PER_HOUR = 60 * exports.MS_PER_MINUTE;
/** @hidden */
exports.TIME_FORMAT = 't';
// probably should calculate this
/** @hidden */
exports.BORDER_WIDTH = 1;
/** @hidden */
exports.DEFAULT_EVENT_HEIGHT = 25;
/** @hidden */
exports.DAYS_IN_WEEK_COUNT = 7;
/** @hidden */
exports.WEEKS_COUNT = 6;
/** @hidden */
exports.HOURS_IN_DAY_COUNT = 24;
/** @hidden */
exports.EMPTY_RESOURCE = [{}];
/** @hidden */
exports.OCCURRENCE_ID = 0;
/** @hidden */
exports.DEFAULT_ID = undefined;
/** @hidden */
exports.DISPLAY_NONE_STYLES = { display: 'none' };
/** @hidden */
exports.VISIBILITY_HIDDEN_STYLES = { visibility: 'hidden' };
/** @hidden */
exports.DEFAULT_GROUP = {
    index: 0,
    resources: []
};
/**
 * @hidden
 */
exports.MIN_DATE = new Date(1900, 0, 1);
/**
 * @hidden
 */
exports.MAX_DATE = new Date(2099, 11, 31);
/** @hidden */
exports.DEFAULT_RANGE = {
    index: 0,
    start: exports.MIN_DATE,
    end: exports.MAX_DATE,
    zonedStart: kendo_date_math_1.ZonedDate.fromLocalDate(exports.MIN_DATE),
    zonedEnd: kendo_date_math_1.ZonedDate.fromLocalDate(exports.MAX_DATE)
};
